content/blog/angular-state-management-with-ngxs/index.md

```diff
@@ -0,0 +1,233 @@
+  styleUrls: ['./app.component.scss']
+})
+export class AppComponent {
+  title = 'ngxs-todo-app';
+
+  @Select(TodoState) todoList$: Observable<ITodo>;
+
+  addForm = new FormGroup({
+    title: new FormControl('', [Validators.required])
+  });
+  constructor(private store: Store){}
+  onSubmit(form: any){
+    this.store.dispatch(new TodoActions.AddTodo(form));
+  }
+  markDone(id: string, is_done: boolean){
+    this.store.dispatch(new TodoActions.markDone(id, is_done));
+  }
+}
+```
+#### 3. Create your actions
+Create folder `state` and put the `todo-actions.ts` file and add below code
+
+```typescript 
+// File name todo-actions.ts
+export class AddTodo {
+    static readonly type = '[Todo] Add';
+    constructor(public payload: any) { }
+}
+
+export class EditTodo {
+    static readonly type = '[Todo] Edit';
+    constructor(public payload: any) { }
+}
+
+export class FetchAllTodos {
+    static readonly type = '[Todo] Fetch All';
+}
+
+export class DeleteTodo {
+    static readonly type = '[Todo] Delete';
+    constructor(public id: number) { }
+}
+```
+
+#### 4. Create your State:
+
+Here we have created the state of the todo, it contails the global state of the todo list.
+Create folder `state` and put the `todo-state.ts` file and add below code
+```typescript
+
+// File name todo-state.ts
+import { Injectable } from '@angular/core';
+import { State, NgxsOnInit, Action, StateContext } from '@ngxs/store';
+import { TodoActions } from './todo-actions';
+import { patch, updateItem } from '@ngxs/store/operators';
+
+export interface ITodo {
+    id: string;
+    title: string;
+    is_done: boolean;
+}
+export interface ITodoStateModel {
+    todoList: ITodo[];
+}
+@State<ITodoStateModel>({
+    name: 'todoList',
+    defaults: {
+        todoList: [],
+    },
+})
+@Injectable()
+export class TodoState implements NgxsOnInit {
+    ngxsOnInit(ctx) {
+        ctx.dispatch(new TodoActions.FetchAllTodos());
+    }
+    @Action(TodoActions.markDone)
+    markDone(
+      ctx: StateContext<ITodoStateModel>, 
+      { payload, is_done }: TodoActions.markDone
+    ) {
+        ctx.setState(
+            patch({
+                todoList: updateItem(
+                  (item: ITodo) => item.id === payload, 
+                  patch({ is_done: !is_done })
+                )
+            })
+        );
+    }
+
+    @Action(TodoActions.AddTodo)
+    add(
+        ctx: StateContext<ITodoStateModel>,
+        { payload }: TodoActions.AddTodo,
+    ) {
+        const state = ctx.getState();
+        ctx.setState({
+            ...state,
+            todoList: [
+                ...state.todoList,
+                {
+                    ...payload,
+                    id: Math.random().toString(36).substring(7),
+                    is_done: false
+                }
+            ],
+        }
+        );
+    }
+
+}
+```
+#### 5. Create your html view
+
+Here we have created a form which we use to create todo and listed all todo.
```

Comment:
This can be `Here we have created a form that we use to create a todo and listed all todos.`